core/content/proxy/content_writer.go

```diff
@@ -76,27 +77,41 @@ func (rw *remoteWriter) Digest() digest.Digest {
 }
 
 func (rw *remoteWriter) Write(p []byte) (n int, err error) {
-	offset := rw.offset
+	const maxBufferSize = defaults.DefaultMaxSendMsgSize >> 1
+	for len(p) > 0 && err == nil {
+		offset := rw.offset
+
+		req := &contentapi.WriteContentRequest{
+			Action: contentapi.WriteAction_WRITE,
+			Offset: offset,
+			Data:   p,
+		}
 
-	resp, err := rw.send(&contentapi.WriteContentRequest{
-		Action: contentapi.WriteAction_WRITE,
-		Offset: offset,
-		Data:   p,
-	})
-	if err != nil {
-		return 0, fmt.Errorf("failed to send write: %w", errgrpc.ToNative(err))
-	}
+		// Limit the size of the data to the maximum buffer size to ensure
+		// the serialized protobuf doesn't exceed the maximum send message size.
+		if len(req.Data) > maxBufferSize {
+			req.Data = req.Data[:maxBufferSize]
+		}
 
-	n = int(resp.Offset - offset)
-	if n < len(p) {
-		err = io.ErrShortWrite
-	}
+		var resp *contentapi.WriteContentResponse
+		resp, err = rw.send(req)
+		if err != nil {
+			return 0, fmt.Errorf("failed to send write: %w", errgrpc.ToNative(err))
+		}
+
+		written := int(resp.Offset - offset)
+		if written < len(req.Data) {
+			err = io.ErrShortWrite
+		}
 
-	rw.offset += int64(n)
-	if resp.Digest != "" {
-		rw.digest = digest.Digest(resp.Digest)
+		rw.offset += int64(written)
+		if resp.Digest != "" {
+			rw.digest = digest.Digest(resp.Digest)
+		}
+		n += written
+		p = p[written:]
```

Comment:
Did you consider [`slices.Chunk`](https://pkg.go.dev/slices@master#Chunk) here? Should make chunk iteration much nicer.

Reply:
I didn't realize this function existed but it also seems to require go 1.23 and the go.mod for this project is on go 1.22.

I can add an implementation using this function for the future though.

Reply:
I believe we target 2 last versions of Go (https://github.com/containerd/containerd/pull/11377), e.g. as of now its 1.24 and 1.23.

`go.mod` directive indicates the _minimum_ Go version that should be used to build the module, this is different from the _required_ Go version.

Reply:
@mxpv I think the go version in go.mod also limit the usage of go features in newer versions?

https://go.dev/doc/modules/gomod-ref#:~:text=The%20go%20directive%20affects%20use%20of%20new%20language%20features:

Reply:
I've rebased and updated the PR to use `slices.Chunk`.